handlers/junebug/junebug.go

```diff
@@ -1,3 +1,227 @@
 package junebug
 
-/* no logs */
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/buger/jsonparser"
+	"github.com/nyaruka/courier"
+	"github.com/nyaruka/courier/handlers"
+	"github.com/nyaruka/courier/utils"
+	"github.com/nyaruka/gocommon/urns"
+	"github.com/pkg/errors"
+)
+
+var (
+	msgLength = 1530
+)
+
+func init() {
+	courier.RegisterHandler(newHandler())
+}
+
+type handler struct {
+	handlers.BaseHandler
+}
+
+func newHandler() courier.ChannelHandler {
+	return &handler{handlers.NewBaseHandler(courier.ChannelType("JN"), "Junebug")}
+}
+
+// Initialize is called by the engine once everything is loaded
+func (h *handler) Initialize(s courier.Server) error {
+	h.SetServer(s)
+	err := s.AddHandlerRoute(h, http.MethodPost, "event", h.ReceiveEvent)
+	if err != nil {
+		return err
+	}
+	return s.AddHandlerRoute(h, http.MethodPost, "inbound", h.ReceiveInbound)
+}
+
+// {
+//   "from": "+27123456789",
+//   "timestamp": "2017-01-01 00:00:00.00",
+//   "content": "content",
+//   "to": "to-addr",
+//   "reply_to": null,
+//   "message_id": "message-id"
+// }
+type moPayload struct {
+	From      string `json:"from"       validate:"required"`
+	Timestamp string `json:"timestamp"  validate:"required"`
+	Content   string `json:"content"`
+	To        string `json:"to"         validate:"required"`
+	ReplyTo   string `json:"reply_to"`
+	MessageID string `json:"message_id" validate:"required"`
+}
+
+// ReceiveInbound is our HTTP handler function for incoming messages
+func (h *handler) ReceiveInbound(ctx context.Context, c courier.Channel, w http.ResponseWriter, r *http.Request) ([]courier.Event, error) {
+	payload := &moPayload{}
+	err := handlers.DecodeAndValidateJSON(payload, r)
+	if err != nil {
+		return nil, courier.WriteAndLogRequestError(ctx, w, r, c, err)
+	}
+
+	// check authentication
+	secret := c.StringConfigForKey(courier.ConfigSecret, "")
+	if secret != "" {
+		authorization := r.Header.Get("HTTP_AUTHORIZATION")
+		if authorization != fmt.Sprintf("Token %s", secret) {
+			return nil, courier.WriteAndLogUnauthorized(ctx, w, r, c, fmt.Errorf("invalid HTTP_AUTHORIZATION header"))
+		}
+	}
+
+	// parse our date
+	date, err := time.Parse("2006-01-02 15:04:05", payload.Timestamp)
+	if err != nil {
+		return nil, courier.WriteAndLogRequestError(ctx, w, r, c, fmt.Errorf("unable to parse date: %s", payload.Timestamp))
+	}
+
+	urn := urns.NewTelURNForCountry(payload.From, c.Country())
+	msg := h.Backend().NewIncomingMsg(c, urn, payload.Content).WithExternalID(payload.MessageID).WithReceivedOn(date.UTC())
+
+	err = h.Backend().WriteMsg(ctx, msg)
+	if err != nil {
+		return nil, err
+	}
+
+	return []courier.Event{msg}, courier.WriteMsgSuccess(ctx, w, r, []courier.Msg{msg})
```

Comment:
@smn one question is whether Junebug only care about status codes when delivering events and messages to RapidPro or whether it also parses the body. Right now this is returning the generic Courier JSON response bodies for incoming msgs and status updates.